### ISSUE TITLE
question: federation secrets best practices?

### ISSUE THREAD

**Issue by jokeyrhyme** [capriza/node-busmq#38]
I'm curious about any recommendations you might have regarding the shared secret used for federation security.

My first thought is that the secret should not be hardcoded in the client code, as anyone with access to the code will be able to see it. Given how easy it often is to extract JavaScript code, I believe this applies equally to both open and closed source client projects.

So, following that, our clients have to request the secret from the web service. It would seem appropriate to require the user to login via other means (e.g. OAuth 2.0, etc), and return the BusMQ secret only to appropriate authenticated users.

Up to this point, I think this covers the basics. Only devices operated by authenticated users are able to access the federated bus.

Going beyond this, I'm considering the implications of starting a separate federated bus for each authenticated user. So the flow would be:

1. client authenticates via HTTPS requests

2. client requests access to a bus and receives the URL and secret

3. client access federated BusMQ endpoint via URL and secret

- client-specific (or user-specific) federated BusMQ endpoint is automatically terminated if there is no activity within a certain threshold (maybe 1 minute? maybe 5 minutes?)

Besides the additional resources required, do you have any comments as to whether this is a good approach, or if this is a bad idea within the current BusMQ design?

**Comment by fujifish** [comment]
You haven't mentioned it explicitly but I gather that you're using the browser client.
Using a static shared secret key in this scenario indeed makes little sense.

You're suggestion for starting a fed server for every client is not such a good idea - It's a separate server per user, listening on a separate port. How would you manage exposing all these ports to the internet? It's both a resource hog as well as a security flaw.

I think a better approach would be to let you specify your own "verify client" method on the fed server instead of specifying a static secret key. That way you can perform any logic you see fit for your use case.

**Comment by jokeyrhyme** [comment]
> You haven't mentioned it explicitly but I gather that you're using the browser client.

Yes, sorry. Forgot to state that.

> You're suggestion for starting a fed server for every client is not such a good idea - It's a separate server per user, listening on a separate port. How would you manage exposing all these ports to the internet?

I could reuse the same HTTP listener in Node.js, with the same port. Just handle a new route per client (or per user) e.g. `/bus/federated/client123` or `/bus/federated/user123`. Unless I'm mistaken and this still consumes a fresh port for each client?

> It's both a resource hog as well as a security flaw.

Definitely uses more resource than a single shared bus. I'm wondering if the CPU and memory usage for n bus federations would be O(log n), O(n), or worst of all O(2^n) or O(n^2)? Unimportant really, as this isn't likely the path I'd take now.

I assume this is a security flaw in terms of denial of service at the very least, where a large number of clients can consume all available ports or all available system resources?

> I think a better approach would be to let you specify your own "verify client" method on the fed server instead of specifying a static secret key. That way you can perform any logic you see fit for your use case.

Ah, I see. I think I found the plumbing for this here:

- https://github.com/capriza/node-busmq/blob/7779b75/lib/fedserver.js#L44
- https://github.com/capriza/node-busmq/blob/7779b75/lib/fedserver.js#L218

I don't have an immediate / urgent requirement to go beyond a single static secret at this stage. But making `verifyClient()` into an option seems like something that could be very useful in future.

**Comment by jokeyrhyme** [comment]
I forgot to add that I still want all clients to be participating in the same pub/sub topics, and all federated buses to ultimately be wired up to the same Redis. This isn't about creating private buses, just thinking through the possibility of creating private access to the same bus.

**Comment by jokeyrhyme** [comment]
Woah! Thanks! This will do nicely. <3

**Comment by fujifish** [comment]
I've added the ability to specify a function instead of a string as the secret for the fedserver.
The function should have the signature of the WebSocketServer.verifyClient function: https://github.com/websockets/ws/blob/master/doc/ws.md#optionsverifyclient

Released version 1.5.4 with this feature.

```javascript
Bus.create({redis: 'http://localhost:6379', federate: {server: fedserver, secret: function(info) {
  // custom logic goes here...
}}});
```

**Comment by jokeyrhyme** [comment]
😺
